In [1]:
import numpy as np
from scipy.stats import fisher_exact, norm

In [2]:
def calc_odds_ratio(a, b, c, d):
    odds_ratio = (a * d) / (b * c)
    p_value = fisher_exact([[a, b], [c, d]])[1]
    se = np.sqrt(1/a + 1/b + 1/c + 1/d)
    z = np.log(odds_ratio) / se
    confint = np.exp(norm.ppf(0.025) * se), np.exp(norm.ppf(0.975) * se)
    return odds_ratio, p_value, confint

In [14]:
#Age
odds_ratio, p_value, confint = calc_odds_ratio(87, 32, 98, 263)

print("The Fisher's Exact Associatio for age =")
print("Odds ratio: ", "{:.3f}".format(odds_ratio))
print("P-value: ", "{:.3e}".format(p_value))
print("confint_lower", "{:.3f}".format(confint[0]))
print("confint_upper", "{:.3f}".format(confint[1]))

The Fisher's Exact Associatio for age =
Odds ratio:  7.296
P-value:  1.179e-18
confint_lower 0.627
confint_upper 1.595


The Fisher's exact odds ratio of 7.296 indicates that the odds of being a positive case in the group with less than 1 year is 7 times higher compared to the group with greater than 1 year.

The p-value of 1.179e-18 is extremely small, which suggests strong evidence against the null hypothesis. In other words, the association between the two groups is statistically significant.

The confidence interval provides a range of values that is likely to contain the true proportion of positive cases with a certain level of confidence, in this case 95% (as defined by the alpha value of 0.05). The lower and upper bounds of the confidence interval are 0.627 and 1.595 respectively, indicating that the true proportion of positive cases in the two groups are likely to be between these values.

In [15]:
#Sex
odds_ratio, p_value, confint = calc_odds_ratio(25, 94, 95, 266)

print("The Fisher's Exact Associatio for sex =")
print("Odds ratio: ", "{:.3f}".format(odds_ratio))
print("P-value: ", "{:.3e}".format(p_value))
print("confint_lower", "{:.3f}".format(confint[0]))
print("confint_upper", "{:.3f}".format(confint[1]))

The Fisher's Exact Associatio for sex =
Odds ratio:  0.745
P-value:  2.733e-01
confint_lower 0.607
confint_upper 1.648
